neuraxle/base.py

```diff
@@ -116,12 +118,85 @@ def append(self, current_id, data_input, expected_output):
         self.expected_outputs.append(expected_output)
 
 
+class BaseSaver(ABC):
+    """
+    Any saver must inherit from this one. Some savers just save parts of objects, some save it all or what remains.
+    """
+
+    @abstractmethod
+    def save_step(self, step: 'BaseStep', context: 'ExecutionContext'):
+        pass  # TODO doc and things
+
+    @abstractmethod
+    def load_step(self, step: 'BaseStep', context: 'ExecutionContext'):
+        pass  # TODO
+
+
+class JoblibPipelineSaver(BaseSaver):
+    """
+    This saver is a good default saver when the object is
+    already stripped out of things that would make it unserializable.
+    """
+    pass  # TODO.
+
+
+class ExecutionContext:
+    def __init__(self, root: str = DEFAULT_CACHE_FOLDER, stripped_saver: JoblibPipelineSaver = None):
+        self.stripped_saver: BaseSaver = stripped_saver
+        self.root: str = root  # TODO: is "./" notation safe on windows? we should be compatible.
+        self.parents: List[BaseStep] = []  # TODO DOC: first is root, second is nested, and so on. This is like a stack.
+
+    def save_all_unsaved(self):
+        # TODO: might move this method elsewhere.
+        while copy(self).pop():
+            pass
+
+        self.peek().save(self)
+
+    def pop_item(self) -> 'BaseStep':
+        # TODO:
+        return self.parents.pop()
+
+    def pop(self) -> bool:
```

Comment:
a pop method shouldn't return bool that doesn't sound like a stack pop method

Reply:
Right. The design or function names are to review here, the thing I like is the fact that we could do a `while pop` thing at the other place in the code so perhaps it needs renaming or something.